Simple audio torch version (#135)
rstudio/ai-blog: _posts/2021-01-29-simple-audio-classification-with-torch/simple-audio-classification-with-torch.Rmd

---
title: "Simple audio classification with torch"
description: > 
  This article attempts to translate the [Daniel Falbel's](https://github.com/dfalbel) ['Simple Audio Classification'](https://blogs.rstudio.com/ai/posts/2018-06-06-simple-audio-classification-keras/) article from `{tensorflow}+{keras}` to `{torch}+{torchaudio}`.
author:
  - name: Athos Damiani
    affiliation: Curso-R
    affiliation_url: https://curso-r.com/
slug: athossimpleaudioclassification

> date: 01-29-2021

Author: OK! I set 'date:' and folder name to 04-02-2021.

      torch::nnf_relu() %>%
      torch::nnf_max_pool2d(kernel_size = c(2,2)) %>%
      
      self$conv4() %>%
      torch::nnf_relu() %>%
      torch::nnf_max_pool2d(kernel_size = c(2,2)) %>%
      
      torch::nnf_dropout(p = 0.25) %>%
      torch::torch_flatten(start_dim = 2) %>%
      
      self$dense1() %>%
      torch::nnf_relu() %>%
      torch::nnf_dropout(p = 0.5) %>%
      self$dense2() 
  }
)

model <- dan_nn()
model$to(device = device)

print(model)
```

## Model fitting

Unlike tensorflow, there is no `model %>% compile(...)` step in torch, so are going to set `loss criterion`, `optimizer strategy` and `evaluation metrics` explicitly in the training loop.

```{r}
loss_criterion <- torch::nn_cross_entropy_loss()
optimizer <- torch::optim_adadelta(model$parameters)
metrics <- list(acc = yardstick::accuracy_vec)
```

### Training loop

```{r}
library(glue)
library(progress)

pred_to_r <- function(x) {
  factor(as.numeric(x$to(device = "cpu")), levels = 1:30)
}

set_progress_bar <- function(total) {
  progress_bar$new(
    total = total, clear = FALSE, width = 70,
    format = ":current/:total [:bar] - :elapsed - loss: :loss - acc: :acc"
  )
}
```

> {torchaudio} comes with 3 loaders, av_loader, tuner_loader, and audiofile_loader - more to come. The `set_audio_backend()` is used to set one of them as the audio loader. Their performances differs from audio format (mp3 or wav). There is no perfect world yet: tuner_loader is the best for mp3, audiofile_loader is the best for wav, but both do not have de option of partial load a sample from audio file without bringing all the data to memory first.

Reviewer: ```suggestion
{torchaudio} comes with 3 loaders, `av_loader`, `tuner_loader`, and `audiofile_loader`- more to come.  `set_audio_backend()` is used to set one of them as the audio loader. Their performances differ based on audio format (mp3 or wav). There is no perfect world yet: `tuner_loader`is best for mp3, `audiofile_loader` is  best for wav, but neither of them has the option of partially loading a sample from an audio file without bringing all the data into memory first.
```

    losses <- c(losses, loss_r)
    
    pb$tick(tokens = list(loss = round(mean(losses), 4), acc = round(mean(accs), 4)))
  })
  
  # test
  predictions_r <- c()
  targets_r <- c()
  coro::loop(for(batch_test in ds_test) {
    predictions <- model(batch_test[[1]]$to(device = device))
    targets <- batch_test[[2]]$to(device = device)
    predictions_r <- c(predictions_r, pred_to_r(predictions$argmax(dim = 2)))
    targets_r <- c(targets_r, pred_to_r(targets))
  })
  val_acc <- metrics$acc(factor(targets_r, levels = 1:30), factor(predictions_r, levels = 1:30))
  cat(glue("val_acc: {val_acc}\n\n"))
}
```

## Making predictions

We already have all predictions calculated for `test_subset`, let's recreate the alluvial plot from original article.

```{r}
library(dplyr)
library(alluvial)
df_validation <- data.frame(
  pred_class = df$classes[predictions_r],
  class = df$classes[targets_r]
)
x <-  df_validation %>%
  mutate(correct = pred_class == class) %>%
  count(pred_class, class, correct)

alluvial(
  x %>% select(class, pred_class),
  freq = x$n,
  col = ifelse(x$correct, "lightblue", "red"),
  border = ifelse(x$correct, "lightblue", "red"),
  alpha = 0.6,
  hide = x$n < 20
)
```

> Model accuracy is `r scales::percent(val_acc, accuracy = 0.01)`, similar to tensorflow version from the original post. Likewise, all conclusions from original post also holds.

Reviewer: ```suggestion
Model accuracy is `r scales::percent(val_acc, accuracy = 0.01)`, similar to the tensorflow version from the original post. Likewise, all conclusions from original post also hold.
```
